### PR TITLE
Implement Cancellable DirectoryReader

### PR DESCRIPTION
**Pull request by matriv**
Implement a Cancellable DirectoryReader that wraps the original
DirectoryReader so that when a search task is cancelled the
DirectoryReaders also stop their work fast. This is usuful for
expensive operations like wilcard/prefix queries where the
DirectoryReaders can spend lots of time and consume resources,
as previously their work wouldn't stop even though the original
search task was cancelled (e.g. because of timeout or dropped client
connection).

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 1]
```diff
@@ -708,6 +708,8 @@ private DefaultSearchContext createSearchContext(SearchRewriteContext rewriteCon
                 searcher, clusterService, indexService, indexShard, bigArrays, threadPool::relativeTimeInMillis, timeout, fetchPhase);
             success = true;
             return searchContext;
+        } catch (IOException e) {
+            return null;
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/SearchService.java]: this leniency looks dangerous?

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java [hunk 2]
```diff
@@ -320,4 +340,99 @@ public DirectoryReader getDirectoryReader() {
...
+
+        @Override
+        public boolean shouldExit() {
+            assert cancellable != null : "checkCancelled must be set immediately after the construction of CancellableIndexReader";
+            if (cancellable.get() == null) {
+                return false;
+            }
+            cancellable.get().run();
+            return false;
+        }
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: This implementation feels a bit awkward, I'd rather like to fork ExitableDirectoryReader entirely to not inherit from its QueryTimeout abstraction.

File: server/src/test/java/org/elasticsearch/search/aggregations/metrics/TDigestPercentileRanksAggregatorTests.java [hunk 3]
```diff
@@ -62,13 +61,19 @@ public void testEmpty() throws IOException {
...
+        try (IndexReader reader = unmappedIndexWriter.getReader()) {
             IndexSearcher searcher = new IndexSearcher(reader);
             PercentileRanks ranks = search(searcher, new MatchAllDocsQuery(), aggBuilder, fieldType);
             Percentile rank = ranks.iterator().next();
             assertEquals(Double.NaN, rank.getPercent(), 0d);
             assertEquals(0.5, rank.getValue(), 0d);
             assertFalse(AggregationInspectionHelper.hasValue(((InternalTDigestPercentileRanks)ranks)));
+        } finally {
+            unmappedIndexWriter.close();
+            directory.close();
```
Comment (jpountz) [inline server/src/test/java/org/elasticsearch/search/aggregations/metrics/TDigestPercentileRanksAggregatorTests.java]: use try-with-resources?

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java [hunk 4]
```diff
@@ -79,8 +89,18 @@
...
-        super(reader);
+    public ContextIndexSearcher(IndexReader reader, Similarity similarity,
+                                QueryCache queryCache, QueryCachingPolicy queryCachingPolicy) throws IOException {
+        this(reader, similarity, queryCache, queryCachingPolicy, true);
+    }
+
+    public ContextIndexSearcher(IndexReader reader, Similarity similarity,
+                                QueryCache queryCache, QueryCachingPolicy queryCachingPolicy,
+                                boolean shouldWrap) throws IOException {
+        super(shouldWrap? new CancellableIndexReader((DirectoryReader) reader, new Cancellable()) : reader);
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: CancellableIndexReader shouldn't have any overhead, so it might be simpler to wrap all the time here.

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java [hunk 5]
```diff
@@ -320,4 +340,99 @@ public DirectoryReader getDirectoryReader() {
         assert reader instanceof DirectoryReader : "expected an instance of DirectoryReader, got " + reader.getClass();
         return (DirectoryReader) reader;
     }
+
+    /**
+     * Wraps an {@link IndexReader} with a cancellation Runnable task.
+     */
+    private static class CancellableIndexReader extends FilterDirectoryReader {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: Let's call it `CancellableDirectoryReader` if it extends DirectoryReader.

**Review by jpountz — COMMENTED** [review]
I left some suggestions but this looks like it's on the right track to me.

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java (hunk 2 above)
Comment (matriv) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: I tried that in the 1st approach but this means we have to copy the whole `ExitablePointValues` to wrap the point values and therefore the `ExitableIntersectVisitor`.

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java (hunk 4 above)
Comment (matriv) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: This boolean was added together with a second constructor because of the AggregatorTestCase and hundreds of tests that derive from that. If we wrap the IndexReader we get:
```
java.lang.AssertionError: The top-reader used to create Weight is not the same as the current reader's top-reader (org.apache.lucene.index.CompositeReaderContext@382edaaa
```
which I tried to fix by changing the AggregatorTestCase to receive IndexReader and not IndexSearcher as an argument. and all the tests to use the derived IndexSearcher from the context created. But even with this there were a few more tests failing that didn't manage to fix, so after discussion with @jimczi we decided to make this workaround for the moment and address the issue in a separate PR afterwards.

I can add a TODO though to not miss it.

File: server/src/main/java/org/elasticsearch/search/SearchService.java (hunk 1 above)
Comment (matriv) [reply server/src/main/java/org/elasticsearch/search/SearchService.java]: Should I simply propagate the exception? or any other suggestion?

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java (hunk 2 above)
Comment (jpountz) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: I don't mind copying it.

**Comment by matriv** [comment]
@jpountz @jimczi I had to fork the ExitableDirectoryReader because of this: https://github.com/elastic/elasticsearch/pull/52822/files#diff-913ad694a7d744ee93ae1dac48d67b0eR415

The CompletionTerms is a final class and the code in CompletionWeight here: https://github.com/apache/lucene-solr/blob/master/lucene/suggest/src/java/org/apache/lucene/search/suggest/document/CompletionWeight.java#L76 expects CompletionTerms, so if they are wrapped into ExitableTerms the if leads to an IllegalArguementException.

File: server/src/main/java/org/elasticsearch/search/query/QueryPhase.java [hunk 6]
```diff
@@ -310,6 +297,8 @@ static boolean executeInternal(SearchContext searchContext) throws QueryPhaseExe
                 queryResult.nodeQueueSize(rExecutor.getCurrentQueueSize());
                 queryResult.serviceTimeEWMA((long) rExecutor.getTaskExecutionEWMA());
             }
+            // Search phase has finished, no longer need to check for timeout
+            searcher.unsetCheckTimeout();
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/query/QueryPhase.java]: Why do we need to do this?

**Review by jpountz — COMMENTED** [review]
Forking sounds totally fine to me.

Can you explain why we need to unset the timeout at some point? I'm guessing that it is to allow aggregations to build their results without errors? A comment would help clarify.

Also I'm not sure why some tests need to become integration tests? These tests used to be integration tests that we migrated to unit test.

File: server/src/main/java/org/elasticsearch/search/query/QueryPhase.java (hunk 6 above)
Comment (matriv) [reply server/src/main/java/org/elasticsearch/search/query/QueryPhase.java]: So that we don't experience a timeout during the aggregation phase. (there were integ tests that were failing because of this). I changed the comment to mention it.

**Comment by matriv** [comment]
> Forking sounds totally fine to me.

Thx!

> Also I'm not sure why some tests need to become integration tests? These tests used to be integration tests that we migrated to unit test.

@jimczi Could you please help explaining?

**Comment by jimczi** [comment]
> Also I'm not sure why some tests need to become integration tests? These tests used to be integration tests that we migrated to unit test.

We can probably fix these tests rather than moving them again into an integration tests. We've made some workaround in the AggregatorTestCase that are now hitting us badly. For instance we allow to use a MultiReader instead of a DirectoryReader so a lot of new tests are using this shortcut even though we always expect a DirectoryReader when used outside of tests. We also have a unit test that simulates the mixing of an unmapped field with a mapped one but the base class `AggregatorTestCase` only exposes a single mapping. I think we should cleanup those tests and ensure that we're making the same assumption in tests than when deploying, however this is out of scope for this pr so I told Marios that we should tackle this in a separate one. To unblock his work we have the workaround here so that we don't try to wrap the reader if it's not a standard directory reader.

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java [hunk 7]
```diff
@@ -320,4 +346,298 @@ public DirectoryReader getDirectoryReader() {
...
+
+        boolean isEnabled();
+        void checkCancelled();
+        default void checkDirReaderCancelled() {
+            checkCancelled();
+        }
+        void unsetCheckTimeout();
+    }
+
+    public static class CancellableImpl implements Cancellable {
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: The implementation should live on the QueryPhase imo. This way you don't need to expose the `checkDirReaderCancelled` and `checkDirReaderCancelled`  in the searcher.

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java [hunk 8]
```diff
@@ -320,4 +346,298 @@ public DirectoryReader getDirectoryReader() {
...
     }
+
+    /**
+     * iFace which implements the query timeout / cancellation logic
+     */
+    public interface Cancellable {
+
+        boolean isEnabled();
+        void checkCancelled();
+        default void checkDirReaderCancelled() {
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: These two functions should not be part of the interface, checkCancelled should be enough

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java [hunk 9]
```diff
@@ -320,4 +346,298 @@ public DirectoryReader getDirectoryReader() {
         assert reader instanceof DirectoryReader : "expected an instance of DirectoryReader, got " + reader.getClass();
         return (DirectoryReader) reader;
     }
+
+    /**
+     * iFace which implements the query timeout / cancellation logic
+     */
+    public interface Cancellable {
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: Can you call QueryCancellable ?

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java [hunk 10]
```diff
@@ -77,10 +89,27 @@
 
     private AggregatedDfs aggregatedDfs;
     private QueryProfiler profiler;
-    private Runnable checkCancelled;
+    private Holder<Cancellable> cancellable;
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: why do you need to use a custom `Holder` ? Now that cancellable directory reader is forked, you should be able to set the cancellable lazily ? So instead of passing the `Cancellable` here you'd set it in `ContextIndexSearcher#setCancellable` ?

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java (hunk 2 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: +1 to copy

File: server/src/main/java/org/elasticsearch/search/query/QueryPhase.java (hunk 6 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/query/QueryPhase.java]: We should do this on the cancellable object directly. The searcher shouldn't be able to unset this way. See my comment on the context searcher.

File: server/src/test/java/org/elasticsearch/search/aggregations/metrics/AvgAggregatorTests.java [hunk 11]
```diff
@@ -611,14 +611,8 @@ public void testCacheAggregation() throws IOException {
         }
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/search/aggregations/metrics/AvgAggregatorTests.java]: Can you revert this change ? It should work without this modification so I'd like to keep this for a different pr since the issue is not related to the exitable directory reader.

File: server/src/test/java/org/elasticsearch/search/aggregations/metrics/MaxAggregatorTests.java [hunk 12]
```diff
@@ -920,14 +920,8 @@ public void testCacheAggregation() throws IOException {
         }
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/search/aggregations/metrics/MaxAggregatorTests.java]: Same here, this shouldn't be needed with the extra constructor.

File: server/src/test/java/org/elasticsearch/search/aggregations/metrics/TDigestPercentileRanksAggregatorTests.java [hunk 13]
```diff
@@ -22,7 +22,6 @@
 import org.apache.lucene.document.Document;
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/search/aggregations/metrics/TDigestPercentileRanksAggregatorTests.java]: Here too ;)

**Review by jimczi — CHANGES_REQUESTED** [review]
It looks good to me overall. I left some comments to simplify the logic in the ContextIndexSearcher and to move the cancellation impl in the QueryPhase directly.

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java (hunk 10 above)
Comment (matriv) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: The issue is that we need a reference to the cancellable (the holder actually) during construction of the Reader: https://github.com/elastic/elasticsearch/pull/52822/files#diff-913ad694a7d744ee93ae1dac48d67b0eR403
So that later on with the `setCancellable` we can set it to the holder.

File: server/src/test/java/org/elasticsearch/search/aggregations/metrics/AvgAggregatorTests.java (hunk 11 above)
Comment (matriv) [reply server/src/test/java/org/elasticsearch/search/aggregations/metrics/AvgAggregatorTests.java]: Sure, will do.

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java [hunk 14]
```diff
@@ -320,4 +346,298 @@ public DirectoryReader getDirectoryReader() {
...
+        private final Cancellable cancellable;
+        private int calls;
+
+        private ExitableIntersectVisitor(PointValues.IntersectVisitor in, Cancellable cancellable) {
+            this.in = in;
+            this.cancellable = cancellable;
+        }
+
+        private void checkAndThrowWithSampling() {
+            if (calls++ % MAX_CALLS_BEFORE_QUERY_TIMEOUT_CHECK == 0) {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: If we want to reduce the overhead to a minimum, we could make `MAX_CALLS_BEFORE_QUERY_TIMEOUT_CHECK` a power of two minus 1 (e.g. `(1 << 4) - 1`, ie. `15`, and use a mask instead of a remainder: `(calls++ & MAX_CALLS_BEFORE_QUERY_TIMEOUT_CHECK) == 0`.

**Review by jpountz — COMMENTED** [review]
Some additional to @jimczi 's.

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java (hunk 14 above)
Comment (matriv) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: Could you please explain why power of two - 1 is better?

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java (hunk 8 above)
Comment (matriv) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: I don't like that much either, I exposed the checkDirReaderCancelled() so that I we can have tests that differentiate between exit in the searcher and exit from the dirReader: https://github.com/elastic/elasticsearch/pull/52822/files#diff-2abce48f3a52657ce3740afade6d5f8fR132
But as discussed I could just unit test the Cancellable reader instead.

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java (hunk 10 above)
Comment (matriv) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: Alreanatively I changed so that the ContextIndexSearcher doesn't know of the holder, but then a cast to `CancellableDirectoryReader` is required to call the setCancellable.

**Comment by matriv** [comment]
@jpountz @jimczi 
Thank you very much for all your help and guidance to tackle this!
I hope I've addressed all comments, please take another look.

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java (hunk 14 above)
Comment (jpountz) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: mostly because masks are typically cheaper than remainders

Comment (matriv) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: Thank you!

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java [hunk 15]
```diff
@@ -244,7 +269,7 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
                 public BulkScorer bulkScorer(LeafReaderContext context) throws IOException {
                     BulkScorer in = weight.bulkScorer(context);
                     if (in != null) {
-                        return new CancellableBulkScorer(in, checkCancelled);
+                        return new CancellableBulkScorer(in, () -> checkCancelled());
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: Let's use a method ref instead in order to avoid creating an anonymous class
```suggestion
                        return new CancellableBulkScorer(in, cancellable::checkCancelled);
```

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java [hunk 16]
```diff
@@ -207,8 +233,7 @@ private void searchLeaf(LeafReaderContext ctx, Weight weight, Collector collecto
             Scorer scorer = weight.scorer(ctx);
             if (scorer != null) {
                 try {
-                    intersectScorerAndBitSet(scorer, liveDocsBitSet, leafCollector,
-                        checkCancelled == null ? () -> { } : checkCancelled);
+                    intersectScorerAndBitSet(scorer, liveDocsBitSet, leafCollector, () -> checkCancelled());
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: can you avoid doing a null check every time we check whether the request should be cancelled, like the code did before?

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java [hunk 17]
```diff
@@ -320,4 +345,266 @@ public DirectoryReader getDirectoryReader() {
...
+    }
+
+    /**
+     * Wraps an {@link IndexReader} with a {@link QueryCancellable}.
+     */
+    static class CancellableDirectoryReader extends FilterDirectoryReader {
+
+        private final Holder<QueryCancellable> cancellableHolder;
+
+        private CancellableDirectoryReader(DirectoryReader in, Holder<QueryCancellable> cancellableHolder) throws IOException {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: this constructor seems to always be called with a `Holder` around a `null`, so let's remove it from the ctor arguments and just set `cancellableHolder = new Holder<>(null);`?

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java [hunk 18]
```diff
@@ -320,4 +345,266 @@ public DirectoryReader getDirectoryReader() {
...
+            return (cancellable.get() != null) ? new ExitablePointValues(pointValues, cancellable.get()) : pointValues;
+        }
+
+        @Override
+        public Terms terms(String field) throws IOException {
+            Terms terms = in.terms(field);
+            if (terms == null) {
+                return null;
+            }
+            return (cancellable.get() != null && terms instanceof CompletionTerms == false) ?
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: can you add a comment about why we're not wrapping CompletionTerms? I also wonder that this `instanceof` logic might be fragile as terms may be wrapped e.g. if document-level security is used?

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java [hunk 19]
```diff
@@ -320,4 +345,266 @@ public DirectoryReader getDirectoryReader() {
...
+        private ExitableTermsEnum(TermsEnum termsEnum, QueryCancellable cancellable) {
+            super(termsEnum);
+            this.cancellable = cancellable;
+            this.cancellable.checkCancelled();
+        }
+
+        @Override
+        public BytesRef next() throws IOException {
+            // Before every iteration, check if the iteration should exit
+            this.cancellable.checkCancelled();
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: Can we do it on a regular interval (like in ExitableIntersectVisitor) instead of for every term, I worry that this might add noticeable overhead.

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java (hunk 17 above)
Comment (matriv) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: The issue is that we need to pass it in the ctor because we need to use it in the constructor: https://github.com/elastic/elasticsearch/pull/52822/files/9695114f3028f98ae59dc5ff7481aab3251bce9f#diff-913ad694a7d744ee93ae1dac48d67b0eR368
so that it's passed down the object tree properly.

Comment (jpountz) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: Ah, I see. Nevermind then.

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java (hunk 18 above)
Comment (matriv) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: Added a comment.

- What other wrapping classes could we encouter?
- Should we have another ExitableTerms class that extends CompletionTerms and also wrap those with cancellation?

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java [hunk 20]
```diff
@@ -208,7 +235,7 @@ private void searchLeaf(LeafReaderContext ctx, Weight weight, Collector collecto
             if (scorer != null) {
                 try {
                     intersectScorerAndBitSet(scorer, liveDocsBitSet, leafCollector,
-                        checkCancelled == null ? () -> { } : checkCancelled);
+                            this.cancellable == null ? () -> {} : this::checkCancelled);
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]:
```suggestion
                            this.cancellable == null ? () -> {} : cancellable::checkCancelled);
```

File: server/src/test/java/org/elasticsearch/search/SearchCancellationTests.java [hunk 21]
```diff
@@ -53,17 +57,19 @@ public static void setup() throws IOException {
         // we need at least 2 segments - so no merges should be allowed
         w.w.getConfig().setMergePolicy(NoMergePolicy.INSTANCE);
         w.setDoRandomForceMerge(false);
-        indexRandomDocuments(w, TestUtil.nextInt(random(), 2, 20));
+        int numDocs = nextInt(random(), 2, 20);
+        indexRandomDocuments(w, numDocs, 0);
         w.flush();
-        indexRandomDocuments(w, TestUtil.nextInt(random(), 1, 20));
+        indexRandomDocuments(w, nextInt(random(), 1, 20), numDocs);
```
Comment (jpountz) [inline server/src/test/java/org/elasticsearch/search/SearchCancellationTests.java]: I tend to prefer fully-qualified calls to `TestUtil#nextInt` instead of adding a static import, given that this method name is quite generic.

File: server/src/test/java/org/elasticsearch/search/SearchCancellationTests.java [hunk 22]
```diff
@@ -92,4 +98,22 @@ public void testCancellableCollector() throws IOException {
...
+        });
+        searcher.getIndexReader().leaves().get(0).reader().terms(STRING_FIELD_NAME).iterator();
+        searcher.getIndexReader().leaves().get(0).reader().getPointValues(POINT_FIELD_NAME).getDocCount();
+
+        cancelled.set(true);
+        expectThrows(TaskCancelledException.class,
+                () -> searcher.getIndexReader().leaves().get(0).reader().terms(STRING_FIELD_NAME).iterator());
+        expectThrows(TaskCancelledException.class,
+                () -> searcher.getIndexReader().leaves().get(0).reader().getPointValues(POINT_FIELD_NAME).getDocCount());
+    }
```
Comment (jpountz) [inline server/src/test/java/org/elasticsearch/search/SearchCancellationTests.java]: let's also test that it can interrupt iteration over the terms/points?

**Review by jpountz — COMMENTED** [review]
Thanks Marios, I left some comments about testing, but this looks almost ready to me.

**Comment by matriv** [comment]
Thank you @jpountz! Please take another look.

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java [hunk 23]
```diff
@@ -320,4 +347,278 @@ public DirectoryReader getDirectoryReader() {
...
+     */
+    public interface QueryCancellable {
+
+        void checkCancelled();
+    }
+
+    /**
+     * Wraps an {@link IndexReader} with a {@link QueryCancellable}.
+     */
+    static class CancellableDirectoryReader extends FilterDirectoryReader {
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: nit: This class deserves its own file. Is there a specific reason to rename from `ExitableDirectoryReader` ?

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java (hunk 17 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: I also wonder if we can replace the `Holder` with a mutable `QueryTimeout` implementation:
Something like:
`````
   private static class MutableQueryTimeout implements QueryTimeout {
        private final Set<Runnable> runnables = new HashSet<>();

        Object add(Runnable action) {
            if (runnables.add(action) == false) {
                throw new IllegalArgumentException("already added");
            }
            return action;
        }

        boolean remove(Runnable action) {
            return runnables.remove(action);
        }

        @Override
        public boolean shouldExit() {
            for (Runnable timeout : runnables) {
                timeout.run();
            }
            return false;
        }

        @Override
        public boolean isTimeoutEnabled() {
            return runnables.size() > 0;
        }
    }
`````
This way you can expose a simple `QueryTimeout` in the wrapped reader and keep the logic simple there ? 
The `ContextIndexSearcher` could then expose the `addQueryTimeout`, `removeQueryTimeout` directly for the `QueryPhase` modification.

File: server/src/main/java/org/elasticsearch/search/query/QueryPhase.java [hunk 24]
```diff
@@ -272,20 +273,8 @@ static boolean executeInternal(SearchContext searchContext) throws QueryPhaseExe
...
-                };
-            } else if (timeoutRunnable != null) {
-                checkCancelled = timeoutRunnable;
-            } else if (cancellationRunnable != null) {
-                checkCancelled = cancellationRunnable;
-            } else {
-                checkCancelled = null;
-            }
-            searcher.setCheckCancelled(checkCancelled);
+            QueryCancellableImpl cancellable = new QueryCancellableImpl(timeoutRunnable, cancellationRunnable);
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/query/QueryPhase.java]: See my comment below, I think this could be simplified into:
`````
            boolean timeoutSet = scrollContext == null && searchContext.timeout() != null &&
                searchContext.timeout().equals(SearchService.NO_TIMEOUT) == false;
            final Runnable timeoutRunnable;
            if (timeoutSet) {
                final long startTime = searchContext.getRelativeTimeInMillis();
                final long timeout = searchContext.timeout().millis();
                final long maxTime = startTime + timeout;
                timeoutRunnable = searcher.addQueryTimeout(() -> {
                    final long time = searchContext.getRelativeTimeInMillis();
                    if (time > maxTime) {
                        throw new TimeExceededException();
                    }
                });
            } else {
                timeoutRunnable = null;
            }

            if (searchContext.lowLevelCancellation()) {
                SearchShardTask task = searchContext.getTask();
                searcher.addQueryTimeout(() -> {
                    if (task.isCancelled()) {
                        throw new TaskCancelledException("cancelled");
                    }
                });
            }
            try {
                ...
            } finally {
              if (timeoutRunnable != null) {
                 searcher.removeQueryTimeout(timeoutRunnable);
              }
           }
`````

**Review by jimczi — COMMENTED** [review]
I think it's getting closer too but we should look at simplifying things because the logic is still hard to follow imo. I made some comments in that direction.

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java (hunk 23 above)
Comment (matriv) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: Only this class or also include the underlying wrapper classes?

Comment (matriv) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: Also if we move it to its own file then it needs to know of the `MutableQueryTimeout`.

Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: > Only this class or also include the underlying wrapper classes?

All of it, yes.

> Also if we move it to its own file then it needs to know of the MutableQueryTimeout.

Why that ? The idea is to have a simple interface in the reader and to use the mutable impl only for the searcher. You can create one in the ctr like this:
````
 public ContextIndexSearcher(IndexReader reader,
                                Similarity similarity,
                                QueryCache queryCache,
                                QueryCachingPolicy queryCachingPolicy) throws IOException {
        this(reader, similarity, queryCache, queryCachingPolicy, new MutableQueryTimeout());
    }

    private ContextIndexSearcher(IndexReader reader,
                                Similarity similarity,
                                QueryCache queryCache,
                                QueryCachingPolicy queryCachingPolicy,
                                MutableQueryTimeout timeout) throws IOException {
        super(ExitableDirectoryReader.wrap((DirectoryReader) reader, timeout));
        this.queryTimeout = timeout;
        setSimilarity(similarity);
        setQueryCache(queryCache);
        setQueryCachingPolicy(queryCachingPolicy);
    }
````
?

**Comment by matriv** [comment]
@jimczi Thank you! Much more elegant approach.

**Review by jimczi — COMMENTED** [review]
Thanks for updating @matriv, I also find it cleaner. I left a comment regarding naming and a possible follow up that I think is important regarding the handling of the Lucene `rewrite`.

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java [hunk 25]
```diff
@@ -77,25 +80,42 @@
...
 
     /**
-     * Set a {@link Runnable} that will be run on a regular basis while
-     * collecting documents.
+     * Add a {@link Runnable} that will be run on a regular basis while
+     * collecting documents and check for query cancellation or timeout.
      */
-    public void setCheckCancelled(Runnable checkCancelled) {
-        this.checkCancelled = checkCancelled;
+    public Runnable addQueryCancellation(Runnable action) {
```
[the remark this answers is not in the document]
Comment (matriv) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: I prefer the `addQueryCancellation` since the cancellation is used not only for the reader.

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java [hunk 26]
```diff
@@ -320,4 +338,34 @@ public DirectoryReader getDirectoryReader() {
         assert reader instanceof DirectoryReader : "expected an instance of DirectoryReader, got " + reader.getClass();
         return (DirectoryReader) reader;
     }
+
+    private static class MutableQueryTimeout implements QueryTimeout {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: Please don't implement QueryTimeout, I think this introduces more issues than it helps, since its contract is that `shouldExit` should return true once the timeout is hit, while we instead use runnables that throw exceptions.

Comment (matriv) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: Replaced with our own iface.

File: server/src/main/java/org/elasticsearch/search/internal/ExitableDirectoryReader.java [hunk 27]
```diff
@@ -0,0 +1,277 @@
...
+ * Wraps an {@link IndexReader} with a {@link QueryCancellation}
+ * which checks for cancelled or timed-out query.
+ */
+class ExitableDirectoryReader extends FilterDirectoryReader {
+
+    interface QueryCancellation {
+
+        boolean isEnabled();
+
+        void checkCancelled();
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/internal/ExitableDirectoryReader.java]: Can you document the contract?

**Review by jpountz — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/search/internal/ExitableDirectoryReader.java (hunk 27 above)
Comment (matriv) [reply server/src/main/java/org/elasticsearch/search/internal/ExitableDirectoryReader.java]: Sure!

**Comment by matriv** [comment]
@jpountz Many thx once again for your patience and help!

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: ok fine with me

File: server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java [hunk 28]
```diff
@@ -77,25 +79,46 @@
...
     }
 
     public void setProfiler(QueryProfiler profiler) {
         this.profiler = profiler;
     }
 
     /**
-     * Set a {@link Runnable} that will be run on a regular basis while
-     * collecting documents.
+     * Add a {@link Runnable} that will be run on a regular basis while fetching document from the
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java]: nit: s/while accessing/while fetching document/ this is confusing since we don't check the cancellation when visiting the stored fields ?

**Review by jimczi — APPROVED** [review]
Thanks for all the iterations @matriv, LGTM!

**Comment by matriv** [comment]
@jimczi Thx you too for you patience and all the help and guidance you provided.
